library/ziti_ctrl.c

```diff
@@ -0,0 +1,293 @@
+
+    struct ctrl_resp *resp = calloc(1, sizeof(struct ctrl_resp));
+    resp->body_parse_func = (void *(*)(const char *, int)) parse_ziti_session;
+    resp->resp_cb = (void (*)(void *, ziti_error*, void *)) cb;
+    resp->ctx = ctx;
+    resp->ctrl = ctrl;
+    resp->ctrl_cb = (void (*)(void *, ziti_error *, struct ctrl_resp *)) ctrl_login_cb;
+
+    req->data = resp;
+}
+
+void ziti_ctrl_logout(ziti_controller *ctrl, void(*cb)(void*, ziti_error*, void*), void *ctx) {
+    um_http_req_t *req = um_http_req(&ctrl->client, "DELETE", "/current-api-session");
+    req->resp_cb = ctrl_resp_cb;
+    req->body_cb = ctrl_body_cb;
+
+    struct ctrl_resp *resp = calloc(1, sizeof(struct ctrl_resp));
+    resp->body_parse_func = NULL; /* no body */
+    resp->resp_cb = (void (*)(void *, ziti_error*, void *)) cb;
+    resp->ctx = ctx;
+    resp->ctrl = ctrl;
+    resp->ctrl_cb = (void (*)(void *, ziti_error *, struct ctrl_resp *)) ctrl_logout_cb;
+
+    req->data = resp;
+}
+
+void ziti_ctrl_get_services(ziti_controller *ctrl, void (*cb)(ziti_service *, ziti_error*, void*), void* ctx) {
+
+    um_http_req_t *req = um_http_req(&ctrl->client, "GET", "/services");
+    req->resp_cb = ctrl_resp_cb;
+    req->body_cb = ctrl_body_cb;
+
+    struct ctrl_resp *resp = calloc(1, sizeof(struct ctrl_resp));
+    resp->body_parse_func = (void *(*)(const char *, int)) parse_ziti_service_array;
+    resp->resp_cb = (void (*)(void *, ziti_error*, void *)) cb;
+    resp->ctx = ctx;
+    resp->ctrl = ctrl;
+    resp->ctrl_cb = (void (*)(void *, ziti_error *, struct ctrl_resp *)) ctrl_service_cb;
+
+    req->data = resp;
+}
+
+void ziti_ctrl_get_service(ziti_controller *ctrl, const char* service_name, void (*cb)(ziti_service *, ziti_error*, void*), void* ctx) {
+    char path[1024];
+    snprintf(path, sizeof(path), "/services?filter=name=\"%s\"", service_name);
+
+    um_http_req_t *req = um_http_req(&ctrl->client, "GET", path);
+    req->resp_cb = ctrl_resp_cb;
+    req->body_cb = ctrl_body_cb;
+
+    struct ctrl_resp *resp = calloc(1, sizeof(struct ctrl_resp));
+    resp->body_parse_func = (void *(*)(const char *, int)) parse_ziti_service_array;
+    resp->resp_cb = (void (*)(void *, ziti_error*, void *)) cb;
+    resp->ctx = ctx;
+    resp->ctrl = ctrl;
+    resp->ctrl_cb = ctrl_service_cb;
+
+    req->data = resp;
+}
+
+void ziti_ctrl_get_net_session(
+        ziti_controller *ctrl, ziti_service *service,
+        void (*cb)(ziti_net_session *, ziti_error*, void*), void* ctx) {
+
+    char *content = malloc(128);
+    size_t len = (size_t) sprintf(content, "{\"serviceId\":\"%s\",\"hosting\":%s}",
+                                      service->id, service->hostable ? "true" : "false");
+
+    um_http_req_t *req = um_http_req(&ctrl->client, "POST", "/sessions");
+    req->resp_cb = ctrl_resp_cb;
+    req->body_cb = ctrl_body_cb;
+    um_http_req_header(req, "Content-Type", "application/json");
+    um_http_req_data(req, content, len, free_body_cb);
```

Comment:
can len be > 128?

Reply:
we're formatting request above, so why would it be >128?